src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumMiniWallet.kt

```diff
@@ -143,19 +159,31 @@ class ElectrumMiniWallet(
     private val job: Job
 
     init {
+
         suspend fun WalletState.processSubscriptionResponse(msg: ScriptHashSubscriptionResponse): WalletState {
             val bitcoinAddress = scriptHashes[msg.scriptHash]
+            val addressMeta = bitcoinAddress?.let { addressMetas[it] }
             return when {
-                bitcoinAddress == null || msg.status.isEmpty() -> this
+                bitcoinAddress == null || msg.status.isEmpty() || addressMeta == null -> this
                 else -> {
                     val unspents = client.getScriptHashUnspents(msg.scriptHash)
-                    val newUtxos = unspents.minus((_walletStateFlow.value.addresses[bitcoinAddress] ?: emptyList()).toSet())
+                    val newUtxos = unspents.minus((_walletStateFlow.value.addresses[bitcoinAddress]?.unspent ?: emptyList()).toSet())
                     // request new parent txs
                     val parentTxs = newUtxos.mapNotNull { utxo -> client.getTx(utxo.txid) }
-                    parentTxs.forEach { tx -> logger.mdcinfo { "received parent transaction with txid=${tx.txid}" } }
-                    val nextWalletState = this.copy(addresses = this.addresses + (bitcoinAddress to unspents), parentTxs = this.parentTxs + parentTxs.associateBy { it.txid })
-                    logger.mdcinfo { "${unspents.size} utxo(s) for address=$bitcoinAddress balance=${nextWalletState.totalBalance}" }
+                    parentTxs.forEach { tx -> logger.info { "received parent transaction with txid=${tx.txid}" } }
+                    val nextAddressState = WalletState.Companion.AddressState(addressMeta, unspents)
+                    val nextWalletState = this.copy(addresses = this.addresses + (bitcoinAddress to nextAddressState), parentTxs = this.parentTxs + parentTxs.associateBy { it.txid })
+                    logger.info { "${unspents.size} utxo(s) for address=$bitcoinAddress index=${addressMeta.indexOrNull ?: "n/a"} balance=${nextWalletState.totalBalance}" }
                     unspents.forEach { logger.debug { "utxo=${it.outPoint.txid}:${it.outPoint.index} amount=${it.value} sat" } }
+                    when (val generator = addressGenerator) {
+                        null -> {}
+                        else -> if (addressMeta.indexOrNull == generator.second - 1 && nextAddressState.unspent.isNotEmpty()) {
+                            logger.info { "requesting generation of next sequence of addresses" }
+                            // if the window = 10 then the first series of address is 0 to 9 and addressGenerator.second = 10
+                            // then when address 9 has utxos, we request generation up until (and excluding) index 10 + 10 = 20, this will generate addresses 10 to 19
+                            mailbox.send(WalletCommand.Companion.GenerateAddress(addressMeta.indexOrNull!! + 1 + generator.first.window))
```

Comment:
Going through the mailbox means that if a user has received funds on 2 different swap-in addresses but in the same block,  there is a (small) chance that the wallet state will be updated twice and result in 2 different channel updates ? Since the generator is local to the wallet it should be possible to avoid since and only update the wallet state once scanning for the next unused addresses has completed ?

Reply:
The `until` below guarantees that we will not call `AddAddress` twice for the same address:
https://github.com/ACINQ/lightning-kmp/blob/16af706680f5b280ae113ce8c38f636f0f690746/src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumMiniWallet.kt#L258-L260

So we will subscribe exactly once to each address and then the wallet state will be updated everytime we receive a `ScriptHashSubscriptionResponse`, I think that's the best we can do?

Reply:
Latest version implements your suggestion: [5c4cbcb](https://github.com/ACINQ/lightning-kmp/pull/584/commits/5c4cbcb5973f65488a5f74854f493c32067bed3f).